packages/amplify-graphql-auth-transformer/src/accesscontrol/acm.ts

```diff
@@ -0,0 +1,140 @@
+import assert from 'assert';
+
+type ACMConfig = {
+  resources: string[],
+  operations: string[]
+}
+
+type SetRoleInput = {
+  role: string,
+  operations: Array<string>,
+  resource?: string,
+};
+
+type ValidateInput = {
+  role?: string,
+  resource?: string,
+  operations?: Array<string>,
+}
+
+type ResourceOperationInput = {
+  operations: Array<string>,
+  role?: string,
+  resource?: string,
+}
+
+/**
+ * Creates an access control matrix
+ * The following vectors are used
+ * - Roles
+ * - Resources
+ * - Operations
+ */
+export class AccessControlMatrix {
+  private roles: Array<string>;
+  private operations: Array<string>;
+  private resources: Array<string>;
+  private matrix: Array<Array<Array<boolean>>>;
+
+  constructor(config: ACMConfig) {
+    this.operations = config.operations;
+    this.resources = config.resources;
+    this.matrix = new Array();
+    this.roles = new Array();
+  }
+
+  public setRole(input: SetRoleInput): void {
+    const { role, resource, operations } = input;
+    this.validate({ resource, operations });
+    let allowedVector: Array<Array<boolean>>;
+    if (!this.roles.includes(role)) {
+      allowedVector = this.getResourceOperationMatrix({ operations, resource });
+      this.roles.push(input.role);
+      this.matrix.push(allowedVector);
+      assert(this.roles.length === this.matrix.length, 'Roles are not aligned with Roles added in Matrix');
+    } else {
+      allowedVector = this.getResourceOperationMatrix({ operations, resource, role });
+      const roleIndex = this.roles.indexOf(role);
+      this.matrix[roleIndex] = allowedVector;
+    }
+  }
+
+  public hasRole(role: string): boolean {
+    return this.roles.includes(role);
+  }
+
+  public isAllowed(role: string, resource: string, operation: string): boolean {
+    this.validate({ role, resource, operations: [operation] });
+    const roleIndex = this.roles.indexOf(role);
+    const resourceIndex = this.resources.indexOf(resource);
+    const operationIndex = this.operations.indexOf(operation);
+    return this.matrix[roleIndex][resourceIndex][operationIndex];
+  }
+
+  public resetAccessForResource(resource: string): void {
+    this.validate({ resource });
+    const resourceIndex = this.resources.indexOf(resource);
+    for(let i=0; i < this.roles.length; i++) {
```

Comment:
```suggestion
    for(let i = 0; i < this.roles.length; i++) {
```

There are a handful of other formatting issues in this file.